### PR TITLE
Test cmd what

### PR DESCRIPTION
**Pull request by runyontr**
**What type of PR is this?**
Add ability to test single functions in `test/cmd/`

**What this PR does / why we need it**:

This PR provides the WHAT argument for running CLI tests.  It also moves the kubeadm tests into the `test/cmd` folder.

For example:

```bash
make test-cmd WHAT="deployment impersonation"
```
calls the `run_deployment_test` and `run_impersonation_test` functions.

```bash
make test-cmd WHAT=kubeadm
```
runs the previosly named `hack/make-rules/test-kubeadm-cmd.sh`

**Which issue(s) this PR fixes**:

None

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
None
```

### CODE REVIEW

File: build/root/Makefile [hunk 1]
```diff
@@ -291,7 +291,7 @@ test-cmd:
 	@echo "$$TEST_CMD_HELP_INFO"
 else
 test-cmd: generated_files
-	hack/make-rules/test-kubeadm-cmd.sh
+	source hack/lib/init.sh
```
Comment (BenTheElder) [inline build/root/Makefile]: Why are we sourcing in the makefile instead of in the script?

File: hack/make-rules/BUILD [hunk 2]
```diff
@@ -41,14 +41,6 @@ sh_binary(
     ],
 )
 
-sh_binary(
```
Comment (BenTheElder) [inline hack/make-rules/BUILD]: This should be moved to the BUILD file at the new location for this script and any references updated.

File: test/cmd/README.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# Kubernetes Command-Line Integration Test Suite
+
+This document describes how Kubernetes should interact with the Kubernetes command-line integration test suite.
```
Comment (BenTheElder) [inline test/cmd/README.md]: I don't think this reads right. I think you mean something like "This document outlines how you can use the command-line integration test-suite"

File: test/cmd/legacy-script.sh [hunk 4]
```diff
@@ -385,468 +386,475 @@ runTests() {
...
+    do 
+      record_command run_${pkg}_tests
+    done
+  else
+    #########################
+    # Kubectl version #
+    #########################
 
-  record_command run_kubectl_version_tests
+    record_command run_kubectl_version_tests
```
Comment (BenTheElder) [inline test/cmd/legacy-script.sh]: Are all these indentation changes necessary?

**Comment by cblecker** [comment]
/uncc
I don't have the review bandwidth for this one

File: hack/make-rules/BUILD (hunk 2 above)
Comment (runyontr) [reply hack/make-rules/BUILD]: I moved this file to `test/cmd/kubeadm.sh`, and is now called by `test-cmd` which remains as part of this file.  I didn't see the other scripts in that file part of this BUILD script, but don't fully understand how Bazel builds/uses this file.

File: build/root/Makefile (hunk 1 above)
Comment (runyontr) [reply build/root/Makefile]: Good point, It looks sourced in `hack/make-rules/test-cmd.sh`, so this is redundant.

File: test/cmd/legacy-script.sh (hunk 4 above)
Comment (runyontr) [reply test/cmd/legacy-script.sh]: This was part of an if/else conditional on the precence of the `WHAT` argument.  I'll rework to escape early and keep the bottom of this with the same formatting

**Comment by spiffxp** [comment]
/sig cli
/sig api-machineruy
/sig cluster-lifecycle
because they all have some ownership of things in cmd

**Comment by spiffxp** [comment]
I'm trying to use locally this to isolate the flake you seem to be running into, and it seems like the local processes don't get torn down at the end?

```
$ make test-cmd WHAT=kubectl_run
# ...
+++ [0222 17:42:07] TESTS PASSED
warning: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely.
No resources found
warning: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely.
No resources found
$ make test-cmd WHAT=kubectl_run
+++ [0222 17:42:28] Running kubectl tests for kube-apiserver

unable to start etcd as port 2379 is in use. please stop the process listening on this port and retry.


tcp4       0      0  127.0.0.1.2379         *.*                    LISTEN

junit report dir: /tmp/junit-results
+++ [0222 17:42:28] Clean up complete
make: *** [test-cmd] Error 1
```

File: hack/make-rules/BUILD (hunk 2 above)
Comment (BenTheElder) [reply hack/make-rules/BUILD]: see the comment at the top of this file, we're _ok_ not to have it as we don't run it this way currently, but it validates what the inputs / dependencies are.

@ixdy thoughts on this? seems like actually maybe we might not need these...

**Comment by spiffxp** [comment]
/retest
pull-kubernetes-e2e-gce failures look like unrelated flakes

File: hack/make-rules/BUILD (hunk 2 above)
Comment (spiffxp) [reply hack/make-rules/BUILD]: I'm pretty sure `test/cmd/BUILD` picks up the renamed file with this:
```
sh_library(
    name = "legacy-script",
    srcs = glob(["*.sh"]),
    data = ["//pkg/kubectl/validation:testdata/v1/validPod.yaml"],
    visibility = ["//visibility:public"],
)
```

And the new script no longer references `hack/ib`

Comment (BenTheElder) [reply hack/make-rules/BUILD]: thanks, resolving

**Comment by spiffxp** [comment]
/approve
I think I might have some whitespace nits but those can be done in a followup, I found this useful as-is

**Comment by spiffxp** [comment]
/lgtm
Yeah, I didn't think I had build/ approve rights
/assign @ixdy @cblecker

**Comment by cblecker** [comment]
/approve
77749a8f-dddb-409b-b63c-2f6f551963e2

**Comment by neolit123** [comment]
```
make test-cmd WHAT=kubeadm
```

the above suggested line builds kubectl and the whole control plane to be able to test this package `cmd/kubeadm/test`. this is completely redundant.

this PR introduces a workflow change without a deprecation period.
right now my quick local fix is to introduce the old `./hack/make-rules/test-kubeadm-cmd.sh`.

**Pull request merged**